src/Common/src/TypeSystem/Common/TypeSystemHelpers.cs

```diff
@@ -36,18 +36,24 @@ static public PointerType MakePointerType(this TypeDesc type)
 
         static public DefType GetClosestDefType(this TypeDesc type)
         {
-            if (type is DefType)
-                return (DefType)type;
-            else
-                return type.BaseType;
+            // The implementation for this should mirror GetClosestMetadataType
+            throw new NotImplementedException();
         }
 
         static public MetadataType GetClosestMetadataType(this TypeDesc type)
         {
-            if (type is MetadataType)
-                return (MetadataType)type;
-            else
-                return type.BaseType.GetClosestMetadataType();
+            if (type.IsSzArray)
+            {
+                MetadataType arrayType = type.Context.SystemModule.GetType("System", "Array`1");
```

Comment:
I would want @davidwrighton opinion on where to put this.

Do we want Array.GetMethods to return the explicitly implemented interface methods? And do we want to tweak virtual/interface resolution logic to accept TypeDesc in a bunch of places where only MetadataType is allowed right now?

Or do we just want to just make GetArrayMetadataType an abstract method on TypeSystemContext that will use this logic to implement an override in CompilerTypeSystemContext?